Entity storage is not very compact

Ahmad-S792 opened the pull request:
#### 6c9eb0e31985f0ab9bcfa34198da9cf10f1dfa6a
<pre>
Entity storage is not very compact

<a href="https://bugs.webkit.org/show_bug.cgi?id=250640">https://bugs.webkit.org/show_bug.cgi?id=250640</a>
rdar://problem/104525785

Reviewed by Darin Adler.

Merge - <a href="https://chromium.googlesource.com/chromium/blink/+/0c7555da9c7b6059b8f55820b5b21685469a42c1">https://chromium.googlesource.com/chromium/blink/+/0c7555da9c7b6059b8f55820b5b21685469a42c1</a>

The entity table contained pointers which both makes it big and prevents
it from being reused as-is from the binary. This changes it to make
the storage more compact by replacing pointers with offsets into a
shared string. Also switching to more compact data types.

The new database will look like:

static const LChar staticEntityStringStorage[] = {
&apos;A&apos;, &apos;E&apos;, &apos;l&apos;, &apos;i&apos;, &apos;g&apos;,
&apos;;&apos;,
&apos;A&apos;, &apos;M&apos;, &apos;P&apos;,
&apos;;&apos;,
&apos;A&apos;, &apos;a&apos;, &apos;c&apos;, &apos;u&apos;, &apos;t&apos;, &apos;e&apos;,
&apos;;&apos;,
...

static const HTMLEntityTableEntry staticEntityTable[2231] = {
    { 0x000C6, 0, 0, 5 }, // &amp;AElig
    { 0x000C6, 0, 0, 6 }, // &amp;AElig;
    { 0x00026, 0, 6, 3 }, // &amp;AMP
    { 0x00026, 0, 6, 4 }, // &amp;AMP;
    { 0x000C1, 0, 10, 6 }, // &amp;Aacute
...

It accomplishes the savings by avoiding pointers in the data (resolving pointers
during startup is expensive), using one string instead of 2000+, not storing the
trailing NUL byte, ordering the members for minimum padding, using minimal size
data types and reusing the same string space for things like &quot;amp;&quot; and &quot;amp&quot;.

This is performance neutral on local testing and does not have any regression on Speedometer
using M1 Pro Macbook.

* Source/WebCore/html/parser/HTMLEntityParser.cpp:
* Source/WebCore/html/parser/HTMLEntitySearch.cpp:
(HTMLEntitySearch::compare):
* Source/WebCore/html/parser/HTMLEntityTable.h:
* Source/WebCore/html/parser/create-html-entity-table:

Canonical link: <a href="https://commits.webkit.org/261245@main">https://commits.webkit.org/261245@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0d273c7ce653c3c69a93fbf55721b0a7be0363e9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/110875 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/20023 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/43471 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/2225 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/119746 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/114836 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/21427 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/11141 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/86/builds/2000 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/116627 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/15915 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/99088 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/103343 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/97875 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/30753 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/44301 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/12550 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/32089 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/86186 "Found 2 new API test failures: /WebKitGTK/TestDownloads:/webkit/Downloads/blob-download, /WebKitGTK/TestGeolocationManager:/webkit/WebKitGeolocationManager/watch-position (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/13107 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/9061 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/18508 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/51710 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7800 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/15052 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/html/parser/HTMLEntityParser.cpp:
```diff
@@ -96,11 +96,12 @@ class HTMLEntityParser {
             for (int i = 0; i < length; ++i) {
                 cc = source.currentCharacter();
-                ASSERT_UNUSED(reference, cc == *reference++);
+                ASSERT_UNUSED(reference, cc == static_cast<UChar>(*reference++));
```
This change doesn’t seem necessary. What happens without it?

darinadler replied:
I also think this assertion is unneeded. We should just remove it entirely. Removing it helps prepare the way for the semicolon technique suggested in the comment below.

Ahmad-S792 replied:
Done and also removed const 'reference' above because it becomes unused.

darinadler commented on Source/WebCore/html/parser/HTMLEntityParser.cpp:
```diff
@@ -96,11 +96,12 @@ class HTMLEntityParser {
             consumedCharacters.clear();
-            const int length = entitySearch.mostRecentMatch()->length;
-            const LChar* reference = entitySearch.mostRecentMatch()->entity;
+            const HTMLEntityTableEntry* mostRecent = entitySearch.mostRecentMatch();
```
I would use `auto*` here.

darinadler replied:
But also, this could be moved up a few more lines. Unclear why we would use a local variable for this here but not just above and below.

Ahmad-S792 replied:
I changed it to auto* right now but will move it above in next push.

Ahmad-S792 replied:
Pushed above at start of function.

darinadler commented on Source/WebCore/html/parser/create-html-entity-table:
```diff
@@ -109,60 +101,126 @@ namespace WebCore {
+def check_ascii(entity_string):
+    for c in entity_string:
+        code = ord(c)
+        assert 0 <= code <= 127, (c + " is not ASCII. Need to change type " +
```
This looks like it might be Google corporate coding style where they follow a strict line length limit. Otherwise we would not break this string over three lines. I suggest we merge this into one longer line.

Ahmad-S792 replied:
Done in one line.

darinadler commented on Source/WebCore/html/parser/HTMLEntityTable.h:
```diff
@@ -29,13 +30,14 @@
-    UChar32 secondValue;
+    UChar secondValue; // UChar since double char sequences only use BMP chars.
+    short entityOffset;
+    short length;
```
Why not `uint8_t`? Maybe it doesn’t help packing, but it seems unlikely we need 15 bits here.

Ahmad-S792 replied:
Done.

darinadler commented on Source/WebCore/html/parser/HTMLEntityTable.h:
```diff
@@ -29,13 +30,14 @@
     UChar32 firstValue;
-    UChar32 secondValue;
+    UChar secondValue; // UChar since double char sequences only use BMP chars.
+    short entityOffset;
```
Would be better to use `uint16_t`.

Ahmad-S792 replied:
Done.

darinadler commented on Source/WebCore/html/parser/create-html-entity-table:
```diff
@@ -109,60 +101,126 @@ namespace WebCore {
+    # and that is NP-Complete or worse.
+    #
+    # This would be even more efficient if we didn't store the
+    # semi-colon in the array but as a bit in the entry.
```
This is a very good point. It would be useful to continue with this optimization and remove the semicolons from the array. I am almost certain this could be done with no performance impact.

Ahmad-S792 replied:
Created - https://bugs.webkit.org/show_bug.cgi?id=253358 to explore in doing it in future. :-)

darinadler commented on Source/WebCore/html/parser/HTMLEntityTable.h:
```diff
@@ -45,6 +47,8 @@ class HTMLEntityTable {
     static const HTMLEntityTableEntry* firstEntryStartingWith(UChar);
     static const HTMLEntityTableEntry* lastEntryStartingWith(UChar);
+
+    static const LChar* entityString(const HTMLEntityTableEntry&);
```
Since this is not null-terminated nor has an intrinsic length I think it is a mistake to use the word string in the function name. These are the characters but they have to be combined with the length above. So I would name this `characters`. It also seems to me it should be a member of `HTMLEntityTableEntry`, not of `HMTLEntityTable`.

Ahmad-S792 replied:
Moved it to `HTMLEntityTableEntry` but has not renamed, will do it in next push. I am testing it in parallel on local branch, hence doing small changes. :-)

Ahmad-S792 replied:
I moved and it led to build failure so I kept it in HTMLEntityTable right now.

Ahmad-S792 replied:
Renaming done as well.

darinadler commented on Source/WebCore/html/parser/HTMLEntitySearch.cpp:
```diff
@@ -45,9 +46,10 @@ HTMLEntitySearch::HTMLEntitySearch()
 HTMLEntitySearch::CompareResult HTMLEntitySearch::compare(const HTMLEntityTableEntry* entry, UChar nextCharacter) const
 {
+    const LChar* entityString = HTMLEntityTable::characters(*entry);
+    UChar entryNextCharacter = entityString[m_currentLength];
```
No, this is not OK. Can’t do this before checking `m_currentLength` against `entry->length`.

Ahmad-S792 replied:
Sure, I will do it. :-)

Ahmad-S792 replied:
Just moved below to original place after `if (entry->length < m_currentLength...` check. Will push next commit, after confirming if you have any other comments or if I have addressed them incorrectly. Thanks!

Ahmad-S792 replied:
Done. :-)

darinadler commented on Source/WebCore/html/parser/HTMLEntityTable.h:
```diff
@@ -45,6 +47,8 @@ class HTMLEntityTable {
     static const HTMLEntityTableEntry* firstEntryStartingWith(UChar);
     static const HTMLEntityTableEntry* lastEntryStartingWith(UChar);
+
+    static const LChar* characters(const HTMLEntityTableEntry&);
```
We can refine further later, and this need not be done now. But I meant this should be a `characters()` non-static member function of `HTMLEntityTableEntry`, which should be possible.

Ahmad-S792 replied:
Created follow-up - https://bugs.webkit.org/show_bug.cgi?id=253360 and if you want me to rename the function to 'entityString', happy to do so. Thanks!

darinadler replied:
I think `characters` is a good name. Should not name it with the word string because it does not return a string.

darinadler commented on Source/WebCore/html/parser/create-html-entity-table:
```diff
@@ -109,60 +101,124 @@ namespace WebCore {
 for entry in entries:
-    output_file.write("static const LChar %s[] = \"%s\";\n" % (convert_entity_to_cpp_name(entry[ENTITY]), entry[ENTITY]))
+    check_ascii(entry[ENTITY])
+    # Reuse substrings from earlier entries. This saves 1-2000
+    # characters, but it's O(n^2) and not very smart. The optimal
+    # solution has to solve the "Shortest Common Superstring" problem
+    # and that is NP-Complete or worse.
```
How long does this take to run? Seeing O(n^2) here makes me slightly worried about build speed.

Ahmad-S792 replied:
Hmm.. Let me do a quick local build with and without these changes and report back. :-)

Ahmad-S792 replied:
WebKit ToT has some build failure errors in WebProcessProxy.h & WebExtensionContext.h and it is 2 AM here, will run it tomorrow morning. :-(

Ahmad-S792 replied:
I tried right now with WebKit ToT ([261219@main](https://commits.webkit.org/261219@main)) with --release flag with and without patch but I think I am doing something wrong and not doing clean build because first with patch, it took 27 minutes (because I refreshed to get all patches till ToT) and then next without patches took less than a minutes. I think my methodology is flawed. Appreciate if you can guide to force clean build.

darinadler replied:
We should merge this. That sounds like too much trouble to me.

Ahmad-S792 replied:
@darinadler - Sorry I was away from laptop. Thanks for adding 'merge-queue'.

Ahmad-S792 committed:
Entity storage is not very compact

https://bugs.webkit.org/show_bug.cgi?id=250640
rdar://problem/104525785

Reviewed by Darin Adler.

Merge - https://chromium.googlesource.com/chromium/blink/+/0c7555da9c7b6059b8f55820b5b21685469a42c1

The entity table contained pointers which both makes it big and prevents
it from being reused as-is from the binary. This changes it to make
the storage more compact by replacing pointers with offsets into a
shared string. Also switching to more compact data types.

The new database will look like:

static const LChar staticEntityStringStorage[] = {
'A', 'E', 'l', 'i', 'g',
';',
'A', 'M', 'P',
';',
'A', 'a', 'c', 'u', 't', 'e',
';',
...

static const HTMLEntityTableEntry staticEntityTable[2231] = {
    { 0x000C6, 0, 0, 5 }, // &AElig
    { 0x000C6, 0, 0, 6 }, // &AElig;
    { 0x00026, 0, 6, 3 }, // &AMP
    { 0x00026, 0, 6, 4 }, // &AMP;
    { 0x000C1, 0, 10, 6 }, // &Aacute
...

It accomplishes the savings by avoiding pointers in the data (resolving pointers
during startup is expensive), using one string instead of 2000+, not storing the
trailing NUL byte, ordering the members for minimum padding, using minimal size
data types and reusing the same string space for things like "amp;" and "amp".

This is performance neutral on local testing and does not have any regression on Speedometer
using M1 Pro Macbook.

* Source/WebCore/html/parser/HTMLEntityParser.cpp:
* Source/WebCore/html/parser/HTMLEntitySearch.cpp:
(HTMLEntitySearch::compare):
* Source/WebCore/html/parser/HTMLEntityTable.h:
* Source/WebCore/html/parser/create-html-entity-table:

Canonical link: https://commits.webkit.org/261245@main
